feat(design): create `DaffTagComponent`

joannalauu opened the pull request:
## PR Checklist

- [x] Commit message follows our [contributing guidelines](https://github.com/graycoreio/daffodil/blob/develop/CONTRIBUTING.md#commit)
- [x] Tests added/updated (for bug fixes/features)
- [x] Documentation added/updated (for bug fixes/features)

## PR Type
<!-- Select the relevant type(s) -->

- [ ] Bug fix
- [x] Feature
- [ ] Style update
- [ ] Refactor
- [ ] Test
- [ ] Build
- [ ] CI
- [ ] Docs
- [ ] Performance
- [ ] Other (please describe)

## Current behavior
Fixes: #3937 

## New behavior
- Added a badge component to the design library
- Badge component takes size, color and status property

## Breaking change?

- [ ] Yes
- [x] No

<!-- If yes, describe impact and migration path -->

## Additional context
<img width="520" height="76" alt="Screenshot 2025-09-23 at 11 17 02 PM" src="https://github.com/user-attachments/assets/6cde852e-925b-45a2-911a-9e38862cf1fd" />
<img width="888" height="115" alt="Screenshot 2025-09-23 at 11 17 07 PM" src="https://github.com/user-attachments/assets/ae8111c4-3a3d-4776-a29c-5617bd0dc502" />
<img width="517" height="70" alt="Screenshot 2025-09-23 at 11 17 12 PM" src="https://github.com/user-attachments/assets/31b3fa03-fdd6-4cc3-afdd-6fbc3c520b8e" />
<img width="900" height="111" alt="Screenshot 2025-09-23 at 11 17 29 PM" src="https://github.com/user-attachments/assets/b54a0319-f1ba-4ee1-ac35-e4d66ac56d7b" />
<img width="486" height="76" alt="Screenshot 2025-09-23 at 11 17 35 PM" src="https://github.com/user-attachments/assets/49a0d4ae-47cb-466c-90fe-0a04efc759c7" />

xelaint commented on libs/design/badge/src/badge.module.ts:
```diff
@@ -0,0 +1,31 @@
+import { CommonModule } from '@angular/common';
```
Since this is a completely new component, the module is unnecessary to create and immediately deprecate. It can be removed.

xelaint commented on libs/design/badge/src/badge/badge.component.html:
```diff
@@ -0,0 +1,3 @@
+<ng-content select="[daffBadgeIcon]"></ng-content>
```
`<ng-content select="[daffBadgeIcon]"></ng-content>` should be replaced with `DaffPrefixabledDirective`.

See the button or notification component for an example on how we do this.

xelaint commented on libs/design/badge/src/badge/badge.component.html:
```diff
@@ -0,0 +1,3 @@
+<ng-content select="[daffBadgeIcon]"></ng-content>
+<ng-content select="[daffBadgeLabel]"></ng-content>
```
The label doesn't need its own directive. This can be:

```
<div class="daff-badge__label">
  <ng-content></ng-content>
</div>
```

xelaint commented on libs/design/badge/src/badge/badge.component.html:
```diff
@@ -0,0 +1,3 @@
+<ng-content select="[daffBadgeIcon]"></ng-content>
+<ng-content select="[daffBadgeLabel]"></ng-content>
+<ng-content select="[daffBadgeDelete]"></ng-content>
```
The delete icon should not be its own component. See `DaffNotificationComponent` for example of the `dismissible` property and how the template should be.

The disabled property can be moved up to the main badge component since the whole badge should be disableable, not just the delete button.

xelaint commented:
Also, sorry for the trouble but can you rename this to `DaffTagComponent`? I think it's a bit more fitting for this component's usage.

joannalauu commented:
@xelaint I have made the requested changes to my pr!! I am not sure why the build pr check failed – is this a problem?

xelaint commented:
> @xelaint I have made the requested changes to my pr!! I am not sure why the build pr check failed – is this a problem?

@sunray4 There's a merge conflict. Needs a rebase against develop

joannalauu committed:
feat(design): create DaffTagComponent

joannalauu commented:
@xelaint I've fixed the issue and added more comprehensive tests for the component. I realized I accidentally made 2 commits for this feature, so I squashed them into 1 to make the commit history more readable. 
Let me know if there are any other changes you want me to add!!

xelaint commented on libs/design/tag/src/tag/tag-sizable.directive.ts:
```diff
@@ -0,0 +1,18 @@
+import { Directive } from '@angular/core';
```
can you fix naming of this file to `tag-sizable.directive.ts`?

xelaint commented on libs/design/tag/src/tag-theme.scss:
```diff
@@ -0,0 +1,139 @@
+		}
+
+		@include dark($mode) {
+			background: daff-color($neutral, 110);
```
```suggestion
			background: daff-color($neutral, 90);
```

xelaint commented on libs/design/tag/src/tag/tag.component.scss:
```diff
@@ -0,0 +1,151 @@
+@use '../../../scss/layout';
+@use '../../../scss/typography' as t;
+
+@mixin daff-tag-sizes {
```
```
@mixin daff-tag-sizes {
	$root: &;

	&.daff-sm {
		font-size: t.$font-size-sm;
		line-height: 1rem;
		gap: 0.375rem;
		padding: 0.5rem;

		#{$root}__close-icon {
			font-size: t.$font-size-sm;
			line-height: 1rem;
		}
	}

	&.daff-md {
		font-size: t.$font-size-base;
		line-height: 1.25rem;
		gap: 0.5rem;
		padding: 0.5rem;

		#{$root}__close-icon {
			font-size: t.$font-size-base;
			line-height: 1.25rem;
		}
	}

	&.daff-lg {
		font-size: t.$font-size-md;
		line-height: 1.5rem;
		gap: 0.75rem;
		padding: 0.75rem;

		#{$root}__close-icon {
			font-size: t.$font-size-md;
			line-height: 1.5rem;
		}
	}
}
```

xelaint commented on libs/design/tag/src/tag/tag.component.scss:
```diff
@@ -0,0 +1,151 @@
+	display: inline-flex;
+	align-items: center;
+	justify-content: center;
+	flex-direction: row;
```
`flex-direction: row;` can be removed.

xelaint commented on libs/design/tag/src/tag/tag.component.scss:
```diff
@@ -0,0 +1,151 @@
+		opacity: 0.5;
+	}
+
+	&__tag-label {
```
```
&__tag-label {
		display: inline-block;
		white-space: nowrap;
		color: currentColor;
	}

	.daff-tag-prefix {
		flex-shrink: 0;
	}
```

can be removed.

xelaint commented on libs/design/tag/src/tag/tag.component.scss:
```diff
@@ -0,0 +1,151 @@
+
+	&__close-icon {
+		@include interactions.clickable();
+		display: inline-flex;
```
```
display: inline-flex;
align-items: center;
justify-content: center;
```

and

```
min-height: 0.875rem;
height: 1rem;
// min-width: 0.66rem;
```

can be removed.

```
appearance: none;` should be added.

xelaint commented on libs/design/tag/src/tag/tag.component.scss:
```diff
@@ -0,0 +1,151 @@
+		height: 1rem;
+		// min-width: 0.66rem;
+
+		fa-icon {
```
`fa-icon` styles can be removed.

xelaint commented on libs/design/tag/src/tag/tag.component.scss:
```diff
@@ -0,0 +1,151 @@
+		}
+	}
+
+	// &__close-icon {
```
remove all commented out code.

xelaint commented on libs/design/tag/src/tag-theme.scss:
```diff
@@ -0,0 +1,139 @@
+	$critical: daff-get-palette($theme, critical);
+	$success: daff-get-palette($theme, success);
+	$neutral: daff-get-palette($theme, neutral);
+	$base: daff-get-base-color($theme, base);
```
```
$base: daff-get-base-color($theme, base);
$base-contrast: daff-get-base-color($theme, base-contrast);
$white: daff-get-base-color($theme, 'white');
$black: daff-get-base-color($theme, 'black');
```

can be removed since these variables are not being used.

xelaint commented on libs/design/tag/src/tag-theme.scss:
```diff
@@ -0,0 +1,139 @@
+			color: daff-color($neutral, 20);
+
+			&.daff-primary {
+				background: daff-color($primary, 70);
```
```suggestion
				background: daff-color($primary, 90);
```

xelaint commented on libs/design/tag/src/tag-theme.scss:
```diff
@@ -0,0 +1,139 @@
+			}
+
+			&.daff-secondary {
+				background: daff-color($secondary, 70);
```
```suggestion
				background: daff-color($secondary, 90);
```

xelaint commented on libs/design/tag/src/tag-theme.scss:
```diff
@@ -0,0 +1,139 @@
+			}
+
+			&.daff-tertiary {
+				background: daff-color($tertiary, 70);
```
```suggestion
				background: daff-color($tertiary, 90);
```

xelaint commented on libs/design/tag/src/tag-theme.scss:
```diff
@@ -0,0 +1,139 @@
+			}
+
+			&.daff-warn {
+				background: daff-color($warn, 70);
```
```suggestion
				background: daff-color($warn, 90);
```

xelaint commented on libs/design/tag/src/tag-theme.scss:
```diff
@@ -0,0 +1,139 @@
+			}
+
+			&.daff-success {
+				background: daff-color($success, 70);
```
```suggestion
				background: daff-color($success, 90);
```

xelaint commented on libs/design/tag/src/tag-theme.scss:
```diff
@@ -0,0 +1,139 @@
+			}
+
+			&.daff-critical {
+				background: daff-color($critical, 70);
```
```suggestion
				background: daff-color($critical, 90);
```

xelaint commented on libs/design/tag/src/tag/tag.component.html:
```diff
@@ -0,0 +1,11 @@
+  <ng-content></ng-content>
+</div>
+@if (dismissible) {
+  <button class="daff-tag__close-icon" (click)="onCloseTag($event)">
```
Add `aria-label="Dismiss"`.

The close button is currently tabbable on a disabled tag. Needs a tabindex of -1 if disabled.

xelaint commented on libs/design/tag/examples/src/basic-tag/basic-tag.component.html:
```diff
@@ -0,0 +1,14 @@
+<daff-tag *ngIf="!hiddenTags['default']" dismissible (closeTag)="hideTag('default')">
```
Only the default example should be displayed here. Move the dismissible and disabled tag examples to their own example components (disabled-tag and dismissible-tag)

xelaint commented on libs/design/tag/examples/src/sizeable-tag/sizeable-tag.component.ts:
```diff
@@ -0,0 +1,41 @@
+  styles: [`
+    :host {
+      display: flex;
+      flex-wrap: wrap;
```
also need `align-items: center;`

xelaint commented on libs/design/tag/examples/src/basic-tag/basic-tag.component.html:
```diff
@@ -0,0 +1,14 @@
+<daff-tag *ngIf="!hiddenTags['default']" dismissible (closeTag)="hideTag('default')">
+  <fa-icon daffPrefix [icon]="faCircleCheck"></fa-icon>
+  <div>Default</div>
```
`<div>` around label not needed in all examples.

joannalauu committed:
feat(design): create DaffTagComponent add requested changes

joannalauu committed:
feat(design): create DaffTagComponent fix linting error

joannalauu commented:
@xelaint I've made the changes that you've requested!! Let me know if there's anything else you'd like me to change.

xelaint commented on libs/design/tag/src/tag/tag.component.ts:
```diff
@@ -0,0 +1,128 @@
+    '[attr.aria-disabled]': 'disabled ? true : null',
+    '[disabled]': 'disabled',
+    '[class.dismissible]': 'dismissible',
+    '[attr.tabindex]': 'disabled ? -1 : this.tabindex',
```
tabindex here can be removed.

xelaint commented on libs/design/tag/src/tag/tag.component.scss:
```diff
@@ -0,0 +1,71 @@
+
+	&[disabled],
+	&.disabled {
+		cursor: not-allowed;
```
The close icon also needs the `cursor: not-allowed;` style when the tag is disabled since it's still showing a pointer on hover.

xelaint commented on libs/design/tag/examples/src/dismissible-tag/dismissible-tag.component.html:
```diff
@@ -0,0 +1,4 @@
+<daff-tag *ngIf="!hiddenTags['dismissible']" dismissible (closeTag)="hideTag('dismissible')">
```
`*ngIf="!hiddenTags['dismissible']"` can be removed.

xelaint commented on libs/design/tag/examples/src/disabled-tag/disabled-tag.component.html:
```diff
@@ -0,0 +1,4 @@
+<daff-tag *ngIf="!hiddenTags['disabled']" dismissible (closeTag)="hideTag('disabled')" disabled>
```
`*ngIf="!hiddenTags['disabled']"` can be removed.

xelaint commented on libs/design/tag/examples/src/colorable-tag/colorable-tag.component.html:
```diff
@@ -0,0 +1,39 @@
+<daff-tag *ngIf="!hiddenTags['default']" dismissible (closeTag)="hideTag('default')">
```
Can you simplify these examples by removing the `dismissible` property?

xelaint commented on libs/design/tag/examples/src/sizeable-tag/sizeable-tag.component.html:
```diff
@@ -0,0 +1,19 @@
+<daff-tag *ngIf="!hiddenTags['default']" dismissible (closeTag)="hideTag('default')">
```
Can you simplify these examples by removing the `dismissible` property?

xelaint commented on libs/design/tag/examples/src/statusable-tag/statusable-tag.component.html:
```diff
@@ -0,0 +1,19 @@
+<daff-tag *ngIf="!hiddenTags['warn']" status="warn" dismissible (closeTag)="hideTag('warn')">
```
Can you simplify these examples by removing the `dismissible` property?

xelaint reviewed (changes requested):
last few changes then it should be gtg! and the `tag-sizeable.directive.ts` file still needs to be renamed.

joannalauu committed:
removed ngif and dimissible property from some examples

joannalauu commented:
@xelaint I've added the changes you've requested!! Is there anything else that I should fix?

xelaint commented on libs/design/tag/src/tag/tag.component.scss:
```diff
@@ -0,0 +1,76 @@
+	border-radius: 0.25rem;
+	position: relative;
+
+	&[disabled],
```
Next the close-icon style within the disabled block here. We prefer a nested structure for combinators for better code clarity!

```
&[disabled],
&.disabled {
	cursor: not-allowed;
	opacity: 0.5;

	#{$root}__close-icon {
		cursor: not-allowed;
	}
}
```

You will need to add `$root: &;` to `.daff-tag` to make this work.

```
.daff-tag {
  $root: &;
  ...
}

joannalauu committed:
nest close icon disabled style

damienwebdev reviewed (approved):
lgtm
